File: web/js/engine/weekly.js
// Port of engine/weekly.py. See that file for why the schedule splits the way
// it does: massed practice before Friday's test, spaced practice after it,
// because the school test and docs/00's success criteria want opposite things
// and both are legitimate.
//
// Dates are handled as ISO strings in UTC throughout, matching schedule.js.
// A local-time Date would shift the test day across a timezone or a DST
// boundary, and "which day is the test" is not a question to get wrong.

import { derive, makeEntry, normalise } from "./derive.js";

export const FRIDAY = 5;            // JS getUTCDay(): Sunday is 0
export const WEEKLY_SHARE = 2 / 3;
export const AFTERGLOW_DAYS = 21;
export const AFTERGLOW_SHARE = 1 / 3;

const DAY = 86400000;
const parseISO = (iso) => Date.parse(`${iso}T00:00:00Z`);
const toISO = (ms) => new Date(ms).toISOString().slice(0, 10);
export const today = () => new Date().toISOString().slice(0, 10);

/** Forgiving on purpose: a school list arrives retyped in a hurry. */
export function parse(text) {
  if (!text) return [];
  const out = [];
  const seen = new Set();
  for (const token of String(text).split(/[^A-Za-z'’-]+/)) {
    const w = token.toLowerCase().replace(/[^a-z]/g, "");
    if (w.length < 3 || seen.has(w)) continue;   // numbering and initials are noise
    seen.add(w);
    out.push(w);
  }
  return out;
}

export function nextTestDay(fromISO, weekday = FRIDAY) {
  const d = fromISO || today();
  const ahead = (weekday - new Date(parseISO(d)).getUTCDay() + 7) % 7;
  return toISO(parseISO(d) + ahead * DAY);
}

export function makeList(text, setOn, testOn, listId) {
  const set = setOn || today();
  return {
    id: listId || `week-${set}`,
    words: parse(text),
    set_on: set,
    test_on: testOn || nextTestDay(set),
    done: false,
  };
}

/**
 * A statutory word keeps its CURATED entry and is merely prioritised; only a
 * genuinely new word gets a derived stub, with what cannot be inferred left
 * empty rather than invented.
 */
export function entriesFor(words, byWord) {
  return words.map((w) => {
    const existing = byWord.get(w);
    return existing ? { ...existing, weekly: true } : makeEntry(w);
  });
}

export function daysUntil(testOn, todayISO) {
  return Math.round((parseISO(testOn) - parseISO(todayISO || today())) / DAY);
}

export function isActive(list, todayISO) {
  if (!list || list.done) return false;
  return daysUntil(list.test_on, todayISO) >= 0;
}

/** Least-seen first, then most-wrong, then list order. */
export function priorityOrder(words, scheduler) {
  return [...words].sort((a, b) => {
    const sa = scheduler.state[a] || { seen: 0, wrong: 0 };
    const sb = scheduler.state[b] || { seen: 0, wrong: 0 };
    return sa.seen - sb.seen || sb.wrong - sa.wrong
        || words.indexOf(a) - words.indexOf(b);
  });
}

function fill(scheduler, picked, size) {
  for (const w of scheduler.session(size * 2)) {
    if (!picked.includes(w)) picked.push(w);
    if (picked.length >= size) break;
  }
  return scheduler.interleave(picked.slice(0, size));
}

/**
 * After the test, keep the week's words on their Leitner schedule for real.
 * Without this they are ten words among a hundred and Scheduler.session()
 * breaks ties alphabetically, so the 1-2-4-8-16 day intervals never land.
 * Only words that are DUE are boosted: this does not re-drill, it lets the
 * interval arrive on the day it says.
 */
function afterglow(scheduler, list, size, todayISO) {
  if (!list) return scheduler.session(size);
  const since = -daysUntil(list.test_on, todayISO);
  if (since < 0 || since > AFTERGLOW_DAYS) return scheduler.session(size);

  const due = new Set(scheduler.due());
  const owed = list.words.filter((w) => due.has(w) && scheduler.state[w]);
  if (!owed.length) return scheduler.session(size);

  const reserve = Math.min(owed.length, Math.max(1, Math.ceil(size * AFTERGLOW_SHARE)));
  return fill(scheduler, priorityOrder(owed, scheduler).slice(0, reserve), size);
}

export function compose(scheduler, list, size = 10, todayISO) {
  const day = todayISO || today();
  if (!isActive(list, day)) return afterglow(scheduler, list, size, day);

  const known = list.words.filter((w) => scheduler.state[w]);
  if (!known.length) return scheduler.session(size);

  const reserve = Math.min(known.length, Math.max(1, Math.ceil(size * WEEKLY_SHARE)));
  return fill(scheduler, priorityOrder(known, scheduler).slice(0, reserve), size);
}

/** Progress for the grown-up view. Counts, never a score for her. */
export function coverage(list, scheduler) {
  const words = list ? list.words : [];
  const st = (w) => scheduler.state[w] || {};
  const practised = words.filter((w) => (st(w).seen || 0) > 0).length;
  return {
    words: words.length,
    practised,
    untouched: words.length - practised,
    secure: words.filter((w) => (st(w).box || 1) >= 3).length,
    days_left: list ? daysUntil(list.test_on) : null,
  };
}

export { normalise, derive };

File: engine/weekly.py
"""
This week's spelling list from school.

THE TENSION THIS MODULE RESOLVES.

Every week the school sends home ten or twenty words and tests them on Friday.
That is a deadline, and a deadline wants massed practice. Everything else in
this app is built on the opposite: spacing beat massing by about 10.6
percentage points across 254 studies, and it is the largest effect the product
rests on.

Both are right, for different horizons, so the schedule serves both:

  BEFORE THE TEST   roughly two thirds of each session is this week's words,
                    one third is statutory words that are due. She gets the
                    list covered without a week-long hole in the spaced review
                    of everything learned before it.

  AFTER THE TEST    the week's words do not disappear. They fold into the
                    Leitner boxes and come back at 1, 2, 4, 8 and 16 days.

That second half is the point. A weekly list that vanishes on Friday can teach
a child to pass Friday, which is not what anyone actually wants: docs/00's
first success criterion is spelling the word correctly in a dictated test A
WEEK OR MORE AFTER practice, and its second is spelling it in her own writing,
which is the KS2 teacher-assessment standard. Folding the list back in is the
only part of this feature that serves either.

IMPLEMENTED AS A COMPOSER, NOT A CHANGE TO THE SCHEDULER. Scheduler is ported
to JavaScript and held to a golden-file replay by two parity tests. Rewriting
its selection would mean regenerating that golden file and re-proving the port.
Composing on top of it leaves both untouched.
"""

import math
import re
from datetime import date, timedelta

import derive

FRIDAY = 4                 # date.weekday(): Monday is 0
WEEKLY_SHARE = 2 / 3       # of each session, before the test
AFTERGLOW_DAYS = 21        # how long a finished list keeps a claim on sessions
AFTERGLOW_SHARE = 1 / 3    # of each session, after the test, for DUE words only


# A line that ends in a colon is a heading, not spellings. Without this, a
# real list pasted verbatim —
#     Noun/verb pairs (N = noun, V = verb):
#     advice (N) / advise (V)
# — put 'noun', 'verb' and 'pairs' on the practice queue, and she would have
# been asked to spell 'pairs'.
HEADING = re.compile(r":\s*$")

# A parenthesised or bracketed tag after a word: '(N)', '(noun)', '[verb]'.
# Schools write the list this way precisely because the word alone is
# ambiguous, so the tag is information, not decoration.
TAG = re.compile(r"[\(\[]\s*([A-Za-z][A-Za-z ]{0,14}?)\s*[\)\]]")
TAG_MEANING = {"n": "noun", "noun": "noun", "v": "verb", "verb": "verb",
               "adj": "adjective", "adjective": "adjective",
               "adv": "adverb", "adverb": "adverb"}


def parse_entries(text):
    """
    Pull words, and any disambiguating tag, out of whatever the adult pastes.

    Returns [{"word": ..., "hint": ... or None}]. Deliberately forgiving: a
    school list arrives retyped in a hurry, numbered, comma-separated, or one
    per row. Headings are skipped; anything else that is not a word is dropped
    rather than queried.
    """
    if not text:
        return []
    out, seen = [], set()
    for line in str(text).splitlines():
        if HEADING.search(line):
            continue
        # Split the line into segments so a tag attaches to the word before it:
        # 'advice (N) / advise (V)' is two items, not one with two tags.
        for segment in re.split(r"[/,;]|\s{2,}", line):
            tag = TAG.search(segment)
            hint = TAG_MEANING.get(tag.group(1).strip().lower()) if tag else None
            for token in re.split(r"[^A-Za-z'’-]+", TAG.sub(" ", segment)):
                w = re.sub(r"[^a-z]", "", token.lower())
                # Numbering and stray initials are noise; real Y5/6 words are 3+.
                if len(w) < 3 or w in seen:
                    continue
                seen.add(w)
                out.append({"word": w, "hint": hint or derive.hint_for(w)})
    return out


def parse(text):
    """Just the words. See parse_entries for the tags."""
    return [e["word"] for e in parse_entries(text)]


def hint_of(lst, word):
    """The tag to speak and show for a word: the list's own first, then the table."""
    w = derive.normalise(word)
    return ((lst or {}).get("hints") or {}).get(w) or derive.hint_for(w)


def next_test_day(from_day=None, weekday=FRIDAY):
    """The next Friday on or after `from_day`."""
    d = from_day or date.today()
    ahead = (weekday - d.weekday()) % 7
    return d + timedelta(days=ahead)


def make_list(text, set_on=None, test_on=None, list_id=None):
    set_on = set_on or date.today()
    entries = parse_entries(text)
    return {
        "id": list_id or f"week-{set_on.isoformat()}",
        "words": [e["word"] for e in entries],
        # word -> 'noun' / 'verb' / ..., for words dictation alone cannot ask
        # for. Only the ones that need it are stored.
        "hints": {e["word"]: e["hint"] for e in entries if e["hint"]},
        "set_on": set_on.isoformat(),
        "test_on": (test_on or next_test_day(set_on)).isoformat(),
        "done": False,
    }


def entries_for(words, by_word):
    """
    An engine entry per word on the list.

    A word already on the statutory list keeps its CURATED entry — full
    origin, morphemes, word family — and is simply prioritised this week. Only
    genuinely new words get a derived one, with the fields that cannot be
    inferred left empty rather than invented. See engine/derive.py.
    """
    out = []
    for w in words:
        existing = by_word.get(w)
        out.append(dict(existing, weekly=True) if existing
                   else derive.make_entry(w))
    return out


def days_until(test_on, today=None):
    today = today or date.today()
    if isinstance(test_on, str):
        test_on = date.fromisoformat(test_on)
    return (test_on - today).days


def is_active(weekly_list, today=None):
    """Before or on test day, and not marked done."""
    if not weekly_list or weekly_list.get("done"):
        return False
    return days_until(weekly_list["test_on"], today) >= 0


def priority_order(words, scheduler):
    """
    Which of this week's words to practise first.

    Least-seen first, then most-wrong, then list order. Front-loading the ones
    she has not met yet is what gets every word covered before Friday; putting
    the ones she keeps missing next is what makes the remaining sessions count.
    """
    def key(w):
        st = scheduler.state.get(w, {"seen": 0, "wrong": 0})
        return (st["seen"], -st["wrong"], words.index(w))
    return sorted(words, key=key)


def compose(scheduler, weekly_list, size=10, today=None):
    """
    The session queue.

    Before the test: weekly words take WEEKLY_SHARE of the slots, the rest
    comes from the ordinary due queue. After it: the ordinary queue alone,
    which by then contains this week's words on their own Leitner schedule.
    """
    today = today or date.today()

    if not is_active(weekly_list, today):
        return _afterglow(scheduler, weekly_list, size, today)

    known = [w for w in weekly_list["words"] if w in scheduler.state]
    if not known:
        return scheduler.session(size=size)

    reserve = min(len(known), max(1, math.ceil(size * WEEKLY_SHARE)))
    picked = priority_order(known, scheduler)[:reserve]

    # Fill the remainder from the ordinary due queue, skipping anything already
    # picked. Ask for extra because some of what comes back will be duplicates.
    if len(picked) < size:
        for w in scheduler.session(size=size * 2):
            if w not in picked:
                picked.append(w)
            if len(picked) >= size:
                break

    return scheduler._interleave(picked[:size])


def _afterglow(scheduler, weekly_list, size, today):
    """
    After the test: keep the week's words on their Leitner schedule, for real.

    Folding them "into the rotation" sounds like it happens by itself, and it
    does not. Once the test passes they are 10 words among 100-odd, and
    Scheduler.session() breaks ties alphabetically, so 'tomorrow' waits behind
    'accommodate' and the 1-2-4-8-16 day intervals never fire on schedule. The
    spacing that is the whole reason for folding them in would be delivered by
    accident of the alphabet or not at all.

    So for three weeks after the test, words from that list that are DUE get
    first call on a third of the session. Due is the operative word: this does
    not re-drill them, it lets the Leitner interval actually land on the day it
    says. After three weeks they are ordinary words with real box positions,
    and the boost is gone.
    """
    if not weekly_list:
        return scheduler.session(size=size)
    since = -days_until(weekly_list["test_on"], today)
    if since < 0 or since > AFTERGLOW_DAYS:
        return scheduler.session(size=size)

    due = set(scheduler.due())
    owed = [w for w in weekly_list["words"] if w in due and w in scheduler.state]
    if not owed:
        return scheduler.session(size=size)

    reserve = min(len(owed), max(1, math.ceil(size * AFTERGLOW_SHARE)))
    picked = priority_order(owed, scheduler)[:reserve]
    for w in scheduler.session(size=size * 2):
        if w not in picked:
            picked.append(w)
        if len(picked) >= size:
            break
    return scheduler._interleave(picked[:size])


def coverage(weekly_list, scheduler):
    """
    How the week is going, for the grown-up view. Counts only, no score for
    her: this is a progress report for an adult, not a target for a child.
    """
    words = weekly_list["words"] if weekly_list else []
    seen = sum(1 for w in words if scheduler.state.get(w, {}).get("seen", 0) > 0)
    secure = sum(1 for w in words if scheduler.state.get(w, {}).get("box", 1) >= 3)
    return {
        "words": len(words),
        "practised": seen,
        "untouched": len(words) - seen,
        "secure": secure,
        "days_left": days_until(weekly_list["test_on"]) if weekly_list else None,
    }
